patches/server/0312-Mob-Spawner-API-Enhancements.patch

```diff
@@ -69,35 +116,189 @@ index 6ba97a0b4f2cb15d5435657c8e8f5c71c6fee3db..c5a117308f051c20b81818ad91e0ca40
++        return type.isEmpty() ? EntityType.PIG : EntityType.fromName(net.minecraft.world.entity.EntityType.getKey(type.get()).getPath());
++    }
++
++    @Override
++    public void setSpawnedType(EntityType creatureType) {
++        if (creatureType == null || creatureType.getName() == null) {
++            throw new IllegalArgumentException("Can't spawn EntityType " + creatureType + " from mobspawners!");
++        }
++
++        getHandle().getSpawner().setEntityId(
++            net.minecraft.world.entity.EntityType.byString(creatureType.getName()).get(),
++            getHandle().getLevel(),
++            getHandle().getLevel().getRandom(),
++            getHandle().blockPosition()
++        );
++    }
++
++    @Override
++    public int getDelay() {
++        return getHandle().getSpawner().spawnDelay;
++    }
++
++    @Override
++    public void setDelay(int delay) {
++        getHandle().getSpawner().spawnDelay = delay;
++    }
++
++    @Override
++    public int getMinSpawnDelay() {
++        return getHandle().getSpawner().minSpawnDelay;
++    }
++
++    @Override
++    public void setMinSpawnDelay(int delay) {
++        getHandle().getSpawner().minSpawnDelay = delay;
++    }
++
++    @Override
++    public int getMaxSpawnDelay() {
++        return getHandle().getSpawner().maxSpawnDelay;
++    }
++
++    @Override
++    public void setMaxSpawnDelay(int delay) {
++        getHandle().getSpawner().maxSpawnDelay = delay;
++    }
++
++    @Override
++    public int getSpawnCount() {
++        return getHandle().getSpawner().spawnCount;
++    }
++
++    @Override
++    public void setSpawnCount(int spawnCount) {
++        getHandle().getSpawner().spawnCount = spawnCount;
++    }
++
++    @Override
++    public int getMaxNearbyEntities() {
++        return getHandle().getSpawner().maxNearbyEntities;
++    }
++
++    @Override
++    public void setMaxNearbyEntities(int maxNearbyEntities) {
++        getHandle().getSpawner().maxNearbyEntities = maxNearbyEntities;
++    }
++
++    @Override
++    public int getRequiredPlayerRange() {
++        return getHandle().getSpawner().requiredPlayerRange;
++    }
++
++    @Override
++    public void setRequiredPlayerRange(int requiredPlayerRange) {
++        getHandle().getSpawner().requiredPlayerRange = requiredPlayerRange;
++    }
++
++    @Override
++    public int getSpawnRange() {
++        return getHandle().getSpawner().spawnRange;
++    }
++
++    @Override
++    public void setSpawnRange(int spawnRange) {
++        getHandle().getSpawner().spawnRange = spawnRange;
++    }
++
++    @Override
++    public net.minecraft.world.level.BaseSpawner getSpawnerHandle() {
++        return getHandle().getSpawner();
++    }
++
++    @Override
++    public net.minecraft.world.level.Level getLevel() {
++        return getHandle().getLevel();
++    }
++
++    @Override
++    public net.minecraft.core.BlockPos getBlockPosition() {
++        return getHandle().blockPosition();
```

Comment:
All of these should be in PaperSpawnerHolder

Reply:
If I move these methods to PaperSpawnerHolder, should I also delete the already exisiting methods in CraftCreatureSpawner? 
Because if I delete them, that will make the diff longer. 
And if I dont remove them, the methods in CraftCreatureSpawner will override the shared implementation in PaperSpawnerHolder. Because of that, the implementation in PaperSpawnerHolder will effectively only be used in CraftMinecartMobSpawner.